src/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -241,12 +248,22 @@ public static IntPtr OffsetOf(Type t, string fieldName)
         /// It must be used with EXTREME CAUTION since passing in an array that is
         /// not pinned or in the fixed heap can cause unexpected results.
         /// </summary>
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        public static extern IntPtr UnsafeAddrOfPinnedArrayElement(Array arr, int index);
+        public static unsafe IntPtr UnsafeAddrOfPinnedArrayElement(Array arr, int index)
+        {
+            if (arr == null)
+                throw new ArgumentNullException(nameof(arr));
 
-        public static IntPtr UnsafeAddrOfPinnedArrayElement<T>(T[] arr, int index)
+            void* pRawData = Unsafe.AsPointer(ref arr.GetRawArrayData());
+            return (IntPtr)((byte*)pRawData + (uint)index * (nuint)arr.GetElementSize());
```

Comment:
Similar checked question. What happens if this overflows? Does the index need to be validated against the array length?

Reply:
This method is documented to not do any bounds checks. If the passed in index is wrong, it can overflow and that's by design.

Reply:
Ok.